File: ProductService/src/main/java/com/ecommercial/shopping/productservice/global/filter/JwtAuthenticationFilter.java
package com.ecommercial.shopping.productservice.global.filter;

import com.ecommercial.shopping.productservice.global.jwt.JwtValidation;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.ServletRequest;
import jakarta.servlet.ServletResponse;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.util.StringUtils;
import org.springframework.web.filter.GenericFilterBean;

import java.io.IOException;

public class JwtAuthenticationFilter extends GenericFilterBean {

    private final JwtValidation jwtValidation;

    public JwtAuthenticationFilter(JwtValidation jwtValidation) {
        this.jwtValidation = jwtValidation;
    }

    @Override
    public void doFilter(
            ServletRequest servletRequest,
            ServletResponse servletResponse,
            FilterChain filterChain
    ) throws IOException, ServletException {
        String token = resolveToken((HttpServletRequest) servletRequest);
        if(token != null && jwtValidation.validateToken(token)) {
            Authentication authentication = jwtValidation.getAuthentication(token);
            SecurityContextHolder.getContext().setAuthentication(authentication);
        }
        filterChain.doFilter(servletRequest,servletResponse);
    }

    public String resolveToken(HttpServletRequest request) {
        String bearerToken = request.getHeader("Authorization");
        if(StringUtils.hasText(bearerToken) &&
                bearerToken.startsWith("Bearer")) {
            return bearerToken.substring(7);
        }
        return null;
    }



}

File: UserService/src/main/kotlin/com/ecommercial/shopping/userservice/global/filter/JwtAuthenticationFilter.kt
package com.ecommercial.shopping.userservice.global.filter

import com.ecommercial.shopping.userservice.global.jwt.JwtProviders
import jakarta.servlet.FilterChain
import jakarta.servlet.GenericFilter
import jakarta.servlet.ServletRequest
import jakarta.servlet.ServletResponse
import jakarta.servlet.http.HttpServletRequest
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.util.StringUtils

class JwtAuthenticationFilter(val jwtProviders: JwtProviders) : GenericFilter() {
    override fun doFilter(
        servletRequest: ServletRequest?,
        servletResponse: ServletResponse?,
        filterChain: FilterChain?
    ) {
        val token: String? = (servletRequest as? HttpServletRequest)?.let { resolveToken(it) }
        println(token)
        if (token != null && jwtProviders.validateToken(token)) {
            val authentication = jwtProviders.getAuthentication(token)
            SecurityContextHolder.getContext().authentication = authentication
        }
        filterChain?.doFilter(servletRequest, servletResponse)
    }

    private fun resolveToken(request: HttpServletRequest): String? {
        val bearToken = request.getHeader("Authorization")
        if (StringUtils.hasText(bearToken) && bearToken.startsWith("Bearer")) {
            return bearToken.substring(7)
        }
        return null
    }
}